sys/signal.c

```diff
@@ -0,0 +1,200 @@
+#include <signal.h>
+#include <thread.h>
+#include <malloc.h>
+#include <stdc.h>
+#include <errno.h>
+#include <sysent.h>
+#include <systm.h>
+#include <proc.h>
+
+static sighandler_t *signal_default_actions[SIG_LAST] = {
+    [SIGINT] = SIG_TERM,  [SIGILL] = SIG_TERM,  [SIGFPE] = SIG_TERM,
+    [SIGABRT] = SIG_TERM, [SIGSEGV] = SIG_TERM, [SIGKILL] = SIG_TERM,
+    [SIGCHLD] = SIG_IGN,  [SIGUSR1] = SIG_TERM, [SIGUSR2] = SIG_TERM,
+};
+
+static const char *signal_names[SIG_LAST] = {
+    [SIGINT] = "SIGINT",   [SIGILL] = "SIGILL",   [SIGABRT] = "SIGABRT",
+    [SIGFPE] = "SIGFPE",   [SIGSEGV] = "SIGSEGV", [SIGKILL] = "SIGKILL",
+    [SIGTERM] = "SIGTERM", [SIGCHLD] = "SIGCHLD", [SIGUSR1] = "SIGUSR1",
+    [SIGUSR2] = "SIGUSR2",
+};
+
+int do_kill(pid_t pid, int sig) {
+  proc_t *target = proc_find(pid);
```

Comment:
I can see a potential problem in here. Suppose you found a process, but after leaving `proc_find` this thread gets preempted and `target` becomes removed in the meantime. I found following code in NetBSD:
```c
/* Find the process */
mutex_enter(proc_lock);
p = proc_find(pid);
if (p == NULL) {
  mutex_exit(proc_lock);
  return ESRCH;
}
mutex_enter(p->p_lock);
mutex_exit(proc_lock);
```
... that makes me believe that `proc_find` should only check if `all_proc_lock` (should be renamed from `all_proc_list_mtx` and made global, apparently) is held when calling it.